src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java

```diff
@@ -0,0 +1,210 @@
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.zip.GZIPInputStream;
+
+import htsjdk.samtools.util.AbstractIterator;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.samtools.util.RuntimeIOException;
+import htsjdk.tribble.readers.AsciiLineReader;
+import htsjdk.tribble.readers.AsciiLineReaderIterator;
+import htsjdk.tribble.readers.PositionalBufferedStream;
+import htsjdk.variant.bcf2.BCF2Codec;
+import htsjdk.variant.bcf2.BCFVersion;
+import htsjdk.variant.variantcontext.VariantContext;
+
+/**
+ * An Class building {@link VCFReaderIterator}
+ * 
+ * Example:
+ * 
+ * <pre>
+ * VCFIterator r = new VCFIteratorBuilder().open(System.in);
+ * while (r.hasNext()) {
+ *     System.out.println(r.next());
+ * }
+ * r.close();
+ * </pre>
+ * 
+ * @author Pierre Lindenbaum / @yokofakun
+ * @see {@link VCFIterator}
+ *
+ */
+
+public class VCFIteratorBuilder {
+
+    /**
+     * creates a VCF iterator from an input stream It detects if the stream is a
+     * BCF stream or a GZipped stream.
+     * 
+     * @param in
+     *            inputstream
+     * @return the VCFIterator
+     * @throws RuntimeIOException
+     */
+    @SuppressWarnings("resource")
+    public VCFIterator open(InputStream in) throws IOException {
+        if (in == null) {
+            throw new IllegalArgumentException("input stream is null");
+            }
+        // try to read the BCF header
+        final BufferedInputStream bufferedinput = new BufferedInputStream(in, 100);
+        bufferedinput.mark(100);
+        final BCFVersion bcfVersion = BCFVersion.readBCFVersion(bufferedinput);
+        bufferedinput.reset();
+
+        //this is VCF or VCF.gz
+        if (bcfVersion == null) {
+            // try to read GZIP header
+            // see http://stackoverflow.com/questions/4818468/how-to-check-if-inputstream-is-gzipped
+            final byte[] signature = new byte[2];
+            bufferedinput.mark(signature.length);
+            final int len = bufferedinput.read(signature); // read the signature
+            bufferedinput.reset(); // push back the signature to the stream
+            if (len == signature.length && signature[0] == (byte) 0x1f && signature[1] == (byte) 0x8b) { 
+                //it's gzip, decode the gzip input
+                in = new GZIPInputStream(bufferedinput);
+            } else {
+                // it's plain VCF
+                in = bufferedinput;
+            }
+            return new VCFReaderIterator(in);
+        } else {
+            return new BCFInputStreamIterator(bufferedinput);
+        }
+    }
+
+    /**
+     * creates a VCF iterator from an URI It detects if the stream is a BCF
+     * stream or a GZipped stream.
+     * 
+     * @param uri
+     *            the URI
+     * @return the VCFIterator
+     * @throws OException
+     */
+    public VCFIterator open(final String uri) throws IOException {
```

Comment:
I think that this method should use `ParsingUtils.openInputStream()` instead, to deal with `java.nio.Path`, `URL` and `File`.